Add AD support to nested solve

dschwen opened the pull request:
Adds a bounded solve capability.

Refs #20247

dschwen committed:
ADify NestedSolve (#20247)

dschwen committed:
Refactor NestedSolve (#20247)

dschwen committed:
Implement bounded solve (#20247)

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/968042/) on 00a61f3 wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/20409/site)

This comment will be updated on new commits.

dschwen committed:
Pass functors by reference! (#20247)

dschwen commented:
Ok, there's another roadblock. The multiplication of two triangular matrices internally creates stack based Eigen matrices ([here](https://github.com/libMesh/libmesh/blob/7ba0889dc15cfa27fbf52ceda6426ff2930ce86b/contrib/eigen/3.3.9/Eigen/src/Core/products/TriangularMatrixMatrix.h#L150)). Those are supposed to be very small - in this case 8x8 - but even that size is too big for the stack when using the big ADReals in the contact module :-(

dschwen commented:
Hm, maybe I can implement a specialization of `Eigen::internal::gebp_traits<Scalar,Scalar>` that sets very small valued for `nr` and `mr`...

recuero commented:
Can that specialization be used depending on the AD container size? I assume that we could have a small performance hit by reducing those values.

dschwen commented:
Yeah, should be possible

dschwen committed:
Add custom ADReal traits for Eigen (#20247)

dschwen commented on modules/tensor_mechanics/test/tests/ad_anisotropic_creep/ad_aniso_creep_temperature_coefficients_function.i:
```diff
@@ -213,7 +213,7 @@
     relative_tolerance = 1e-20
     absolute_tolerance = 1e-20
-    internal_solve_output_on = never
+    # internal_solve_output_on = never
```
Yes, currently we do. but with the restructuring I did in this PR we could recover that info.

dschwen commented on framework/include/utils/NestedSolve.h:
```diff
@@ -247,43 +211,68 @@ NestedSolve::nonlinear(Real & guess, R computeResidual, J computeJacobian)
+      else
+        guess_eigen(0) = bounds.second;
+
+      if (bounds_hit)
+      {
+        _state = State::CONVERGED_BOUNDS;
+        return;
+      }
+
+      bounds_hit = true;
     }
```
> // Snap to bounds. Terminate solve if we snap twice consecutively.

dschwen committed:
Add test, address comment (#20247)

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1116337/) on 6dfc5cd wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/20409/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/20409/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format b2ec4f96f9c10b73c4cb5407c96b25ac76319043`

moosebuild commented on large_media:
```diff
@@ -1 +1 @@
-Subproject commit 937327ae580cea63ce6ff0e4cc5b9d64160f48e0
+Subproject commit e1211aaf2dec766e9a8e2da1631ee2f36ec7c5de
```
**Caution!** This contains a submodule update

dschwen commented:
Is there something more you need me to do on this PR @recuero ?

dschwen committed:
Fix AD support (#20247)

dschwen committed:
Emulate old convergence tests (#20247)

dschwen committed:
Use partial pivot LU by default

chaibhave committed:
Added moose enum to check for x tolerance (#20247)

of delta variable in nonlinear method. Added a
return value for compute lamda function to allow
damping the linear iteration if desired.

dschwen committed:
Fix x-tol, remove dampening for now (#20247)

dschwen commented on framework/include/utils/NestedSolve.h:
```diff
@@ -318,40 +420,290 @@ NestedSolve::nonlinear(V & guess, T compute)
     linear(jacobian, delta, residual);
+
+    // Check if step size is smaller than the floating point tolerance
+    if (normSquare(delta) <= _x_tolerance_square)
```
@ChaitanyaBhave-e this is not the same as the componentwise abs, but this method has overloads for every value type.

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1420496/) on 7672b19 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/20409/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/20409/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format a19b47faac73047a692a8abb8a4b508b9f1a085d`

dschwen commented:
Keep

tophmatthews commented:
+1 to this...would like to use bounds in `NestedSolve`

dschwen commented:
I'm reducing the scope of this PR and will revisit the use if the class in anisotropic return mapping at a later time. I have the anisotropic elastoplasticity tests mostly working except for three CSV diffs, which seem to be tolerance related (resulting is slightly different material time steps).

dschwen committed:
Use nested solve in newton material (#20247)

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/1774980/) on 64366ab wanted to post the following:


The following coverage requirement(s) failed:
- Failed to generate `framework` coverage rate (required: 81.0%)

GiudGiud commented on framework/include/utils/EigenADReal.h:
```diff
@@ -0,0 +1,117 @@
+    RhsPacketSize = 1,
+    ResPacketSize = 1,
+    NumberOfRegisters = 1,
+    nr = 1,
+    mr = 1,
```
wrt to running out of memory we're safe.
Does Eigen give you better defaults though? Looks like Core has some architecture dependent logic on the registers for example

GiudGiud replied:
should there be an example of setting these right in the unit tests?

dschwen replied:
you don't need to set those

GiudGiud commented on framework/src/utils/NestedSolve.C:
```diff
@@ -11,66 +11,94 @@
+                        absoluteToleranceDefault(),
                         "Absolute convergence tolerance for Newton iteration");
+  params.addParam<Real>(
+      "x_tolerance", xToleranceDefault(), "Threshold for minimum step size of linear iterations");
```
why x? x for linear?

dschwen replied:
that seems to be a fairly common term. it is `x` as in `f(x)=0`, this is the tolerance for changes in the solution. If the solution change is less than this value we assume the iterations to be converged.

GiudGiud replied:
x = f(x) is the nonlinear problem ?

GiudGiud commented on framework/include/utils/NestedSolve.h:
```diff
@@ -318,40 +420,290 @@ NestedSolve::nonlinear(V & guess, T compute)
+ * Adaptor functor for Eigen::Matrix based residual and Jacobian types. No type conversion
+ * required.
+ */
+template <bool is_ad, typename R, typename J, bool store_redidual_norm>
```
```suggestion
template <bool is_ad, typename R, typename J, bool store_residual_norm>
```

GiudGiud commented on framework/include/utils/NestedSolve.h:
```diff
@@ -318,40 +420,290 @@ NestedSolve::nonlinear(V & guess, T compute)
+  int operator()(V & guess, V & residual)
+  {
+    _residual_lambda(guess, residual);
+    if constexpr (store_redidual_norm)
```
```suggestion
    if constexpr (store_residual_norm)
```

dschwen commented on framework/include/utils/EigenADReal.h:
```diff
@@ -0,0 +1,119 @@
+struct gebp_kernel<ADReal, ADReal, Index, DataMapper, 1, 1, ConjugateLhs, ConjugateRhs>
+{
+  EIGEN_DONT_INLINE
+  void operator()(const DataMapper & res,
```
where are we modifying it?

GiudGiud replied:
res(i,j) += ?

dschwen replied:
That has nothing to do with this. This is an Eigen class that describes how to use the ADReal datatype

dschwen commented on unit/src/NestedSolveTest.C:
```diff
@@ -162,6 +186,20 @@ TEST(NestedSolve, ScalarPowell)
   EXPECT_NEAR(solution, 2, 1e-6);
 }
 
+TEST(NestedSolve, ScalarPowellTight)
```
Powell's dog leg method

GiudGiud replied:
pretty cool

dschwen commented on test/src/materials/NewtonMaterial.C:
```diff
@@ -54,12 +53,18 @@ NewtonMaterial::computeQpProperties()
+  // success state if the initial guess was exact).
+  if (getParam<unsigned int>("max_iterations") > 0)
+    _nested_solve.nonlinear(_p[_qp],
+                            // Functor to compute residual and jacobian. The initial guess is not
```
_Functor_ is the right term here! Thic does not have to be a _Lambda_, it could be a class with an overloaded `operator()` for example. I'd be happy to change it to `C++ functor (e.g. lambda)` to avoid confusion with MOOSE functors.

dschwen replied:
Oh, it is a lambda here, so nevermind

dschwen commented on framework/include/utils/NestedSolve.h:
```diff
@@ -90,152 +151,167 @@ class NestedSolve
-  class EigenAdaptorFunctor
+  // compute first residual norm for relative convergence checks
+  auto r0_square = r_square;
+  if (r0_square == 0)
```
`State::EXACT_GUESS` implies exact, otherwise we iterate. I suppose we could alternatively check for the absolute tolerance. :-/

dschwen committed:
Address comments (#20247)

dschwen commented on framework/include/utils/NestedSolve.h:
```diff
@@ -297,9 +348,60 @@ NestedSolve::nonlinear(RealVectorValue & guess, R computeResidual, J computeJaco
+  guess_eigen << guess;
+
+  auto status = solver.solveInit(guess_eigen);
+  if (status == Eigen::HybridNonLinearSolverSpace::ImproperInputParameters)
+    return;
```
I'll do a follow on PR for this
